src/main/java/org/fiware/iam/dsp/RainbowQuoteHandler.java

```diff
@@ -0,0 +1,166 @@
+                                .map(t -> HttpResponse.noContent()));
+    }
+
+    @Override
+    public Mono<HttpResponse<?>> handleQuoteStateChange(QuoteVO quoteVO) {
+        QuoteStateTypeVO quoteStateTypeVO = quoteVO.getState();
+        log.debug("Quote state is {}", quoteStateTypeVO);
+        return switch (quoteStateTypeVO) {
+            case APPROVED ->
+                    rainbowAdapter.updateNegotiationProcessByProviderId(quoteVO.getExternalId(), "dspace:OFFERED")
+                            .map(t -> HttpResponse.noContent());
+            case ACCEPTED -> tmForumAdapter.getConsumerDid(quoteVO)
+                    .flatMap(consumerDid ->
+                            rainbowAdapter
+                                    .createAgreementAfterNegotiation(quoteVO.getExternalId(), consumerDid, generalProperties.getDid())
+                                    .flatMap(r -> rainbowAdapter.updateNegotiationProcessByProviderId(quoteVO.getExternalId(), "dspace:AGREED")
+                                            .map(t -> HttpResponse.noContent())));
+            // a lot of requests can just be ignored
+            default -> Mono.just(HttpResponse.noContent());
+        };
+    }
+
+    @Override
+    public Mono<HttpResponse<?>> handleQuoteDeletion(QuoteVO quoteVO) {
+        if (quoteVO.getExternalId() == null && !quoteVO.getExternalId().isEmpty()) {
+            return rainbowAdapter.updateNegotiationProcessByProviderId(quoteVO.getExternalId(), "dspace:TERMINATED")
+                    .map(t -> HttpResponse.noContent());
+        }
+
+        return Mono.just(HttpResponse.noContent());
+    }
+
+    private String getOfferIdFromQuote(QuoteVO quoteVO) {
+        return getRelevantQuoteItems(quoteVO)
+                .map(QuoteItemVO::getProductOffering)
+                .map(org.fiware.iam.tmforum.quote.model.ProductOfferingRefVO::getId)
+                .findFirst()
+                .orElseThrow(() -> new IllegalArgumentException("The event does not reference an offer."));
+    }
+
+    private Stream<QuoteItemVO> getRelevantQuoteItems(QuoteVO quoteVO) {
+        return quoteVO
+                .getQuoteItem()
+                .stream()
+                .filter(qi -> !qi.getState().equals("rejected"));
+    }
+
+    private List<Policy> getPoliciesFromQuote(ObjectMapper objectMapper, QuoteVO quoteVO) {
+        return getRelevantQuoteItems(quoteVO)
+                .map(QuoteItemVO::getUnknownProperties)
+                .map(Map::entrySet)
+                .flatMap(Set::stream)
+                .filter(ap -> ap.getKey().equals("policy"))
+                .map(Map.Entry::getValue)
+                .filter(List.class::isInstance)
+                .map(List.class::cast)
+                .flatMap(List::stream)
+                .map(v -> objectMapper.convertValue(v, Policy.class))
+                .toList();
+    }
+
+    private List<PermissionVO> getPermissionsFromQuote(ObjectMapper objectMapper, QuoteVO quoteVO) {
```

Comment:
Why do you hand in a objectmapper? As far as I see its only used in the handleQuoteCreation method and uses the class's mapper

Reply:
Left over, removed it